sbot/power_board.py

```diff
@@ -0,0 +1,518 @@
+from __future__ import annotations
+
+import atexit
+import logging
+from enum import IntEnum
+from types import MappingProxyType
+
+from serial.tools.list_ports import comports
+
+from .exceptions import BoardDisconnectionError, IncorrectBoardError
+from .logging import log_to_debug
+from .serial_wrapper import SerialWrapper
+from .utils import Board, BoardIdentity, float_bounds_check, get_USB_identity
+
+logger = logging.getLogger(__name__)
+
+
+class PowerOutputPosition(IntEnum):
+    """
+    A mapping of output name to number of the PowerBoard outputs.
+
+    The numbers here are the same as used in communication protocol with the PowerBoard.
+    """
+    H0 = 0
+    H1 = 1
+    L0 = 2
+    L1 = 3
+    L2 = 4
+    L3 = 5
+    FIVE_VOLT = 6
+
+
+# This output is always on, and cannot be controlled via the API.
+BRAIN_OUTPUT = PowerOutputPosition.FIVE_VOLT
+
+
+class PowerBoard(Board):
+    """
+    A class representing the power board interface.
+
+    This class is intended to be used to communicate with the power board over serial.
+
+    :param serial_port: The serial port to connect to.
+    :param initial_identity: The identity of the board, as reported by the USB descriptor.
+    """
+    __slots__ = (
+        '_serial', '_identity', '_outputs', '_battery_sensor',
+        '_piezo', '_run_led', '_error_led')
+
+    @staticmethod
+    def get_board_type() -> str:
+        """
+        Return the type of the board.
+
+        :return: The literal string 'PBv4B'.
+        """
+        return 'PBv4B'
+
+    def __init__(
+        self,
+        serial_port: str,
+        initial_identity: BoardIdentity | None = None,
+    ) -> None:
+        if initial_identity is None:
+            initial_identity = BoardIdentity()
+        self._serial = SerialWrapper(serial_port, 115200, identity=initial_identity)
+
+        self._outputs = Outputs(self._serial)
+        self._battery_sensor = BatterySensor(self._serial)
+        self._piezo = Piezo(self._serial)
+        self._run_led = Led(self._serial, 'RUN')
+        self._error_led = Led(self._serial, 'ERR')
+
+        self._identity = self.identify()
+        if self._identity.board_type != self.get_board_type():
+            raise IncorrectBoardError(self._identity.board_type, self.get_board_type())
+        self._serial.set_identity(self._identity)
+
+        atexit.register(self._cleanup)
+
+    @classmethod
+    def _get_supported_boards(
+        cls, manual_boards: list[str] | None = None,
+    ) -> MappingProxyType[str, PowerBoard]:
+        """
+        Find all connected power boards.
+
+        Ports are filtered to the USB vendor and product ID: 0x1BDA and 0x0010 respectively.
+
+        :param manual_boards: A list of manually specified serial ports to also attempt
+            to connect to, defaults to None
+        """
+        boards = {}
+        serial_ports = comports()
+        for port in serial_ports:
+            if port.vid == 0x1BDA and port.pid == 0x0010:
```

Comment:
Same as motor board, do we need comment/variable to say what these are?

Reply:
05a96aae1e73a3fe1a7ff92f3360334b24f3095e